system-helper/flatpak-system-helper.c

```diff
@@ -51,6 +59,236 @@ G_DEFINE_AUTOPTR_CLEANUP_FUNC (AutoPolkitAuthorizationResult, g_object_unref)
 G_DEFINE_AUTOPTR_CLEANUP_FUNC (AutoPolkitDetails, g_object_unref)
 G_DEFINE_AUTOPTR_CLEANUP_FUNC (AutoPolkitSubject, g_object_unref)
 
+typedef struct
+{
+  FlatpakSystemHelper *object;
+  GDBusMethodInvocation *invocation;
+  GCancellable *cancellable;
+
+  guint watch_id;
+  uid_t uid;
+
+  gint ref_count;
+
+  gchar *ref;
+  gchar *commit;
+  gchar *root_dir_name;
+  gchar *user_dir_name;
+  gchar *child_repo_path;
+
+  GUnixMountMonitor *mount_monitor;
+  gulong mount_monitor_id;
+  guint unsuccessful_mount_timeout_id;
+  GSubprocess *bindfs;
+} FlatpakSystemHelperOngoingPull;
+
+static void
+system_helper_ongoing_pull_free (FlatpakSystemHelperOngoingPull *pull)
+{
+  g_autoptr(GFile) root_dir = NULL;
+  g_autoptr(GFile) user_dir = NULL;
+  g_autoptr(GError) local_error = NULL;
+
+  G_LOCK (ongoing_pulls);
+  g_ptr_array_remove (ongoing_pulls, pull);
+  G_UNLOCK (ongoing_pulls);
+
+  root_dir = g_file_new_for_path (pull->root_dir_name);
+  user_dir = g_file_new_for_path (pull->user_dir_name);
+
+  g_cancellable_cancel (pull->cancellable);
+
+  g_clear_object (&pull->bindfs);
+
+  if (!flatpak_rm_rf (root_dir, NULL, &local_error))
+    {
+      g_warning ("Unable to remove ongoing pull's root dir at: %s", local_error->message);
+      g_clear_error (&local_error);
+    }
+
+  if (!flatpak_rm_rf (user_dir, NULL, &local_error))
+    {
+      g_warning ("Unable to remove ongoing pull's user dir at: %s", local_error->message);
+      g_clear_error (&local_error);
+    }
+
+  g_clear_object (&pull->mount_monitor);
+
+  g_clear_pointer (&pull->root_dir_name, g_free);
+  g_clear_pointer (&pull->user_dir_name, g_free);
+  g_clear_pointer (&pull->child_repo_path, g_free);
+  g_clear_pointer (&pull->ref, g_free);
+  g_clear_pointer (&pull->commit, g_free);
+
+  g_slice_free (FlatpakSystemHelperOngoingPull, pull);
+}
+
+static void
+system_helper_ongoing_pull_teardown (FlatpakSystemHelperOngoingPull *pull)
+{
+  if (pull->watch_id)
+    g_clear_handle_id (&pull->watch_id, g_source_remove);
```

Comment:
The clear function here should be `g_bus_unwatch_name`, not `g_source_remove`. Bus name watches are not `GSource`s.